server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java

```diff
@@ -337,6 +337,10 @@ private void extractItem(String path, XContentParser source) throws IOException
                     case VALUE_NULL:
                         source.nextToken();
                         break;
+                    case VALUE_NUMBER: // allow parsing numbers assuming routing fields are always keyword fields
```

Comment:
Maybe also add a unit test for this in `IndexRoutingTests`?

Reply:
As we discussed in our sync, let's add a test that makes use of `assertIndexShard(...)` and provides a source that contains a number for the routing field.